cpp/src/arrow/array/builder_dict.cc

```diff
@@ -30,11 +33,117 @@
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/hashing.h"
 #include "arrow/util/logging.h"
+#include "arrow/visitor_inline.h"
 
 namespace arrow {
 
 using internal::checked_cast;
 
+// ----------------------------------------------------------------------
+// DictionaryType unification
+
+struct UnifyDictionaryValues {
+  MemoryPool* pool_;
+  std::shared_ptr<DataType> value_type_;
+  const std::vector<const DictionaryType*>& types_;
+  std::shared_ptr<Array>* out_values_;
+  std::vector<std::vector<int32_t>>* out_transpose_maps_;
+
+  Status Visit(const DataType&, void* = nullptr) {
+    // Default implementation for non-dictionary-supported datatypes
+    std::stringstream ss;
+    ss << "Unification of " << value_type_->ToString()
+       << " dictionaries is not implemented";
+    return Status::NotImplemented(ss.str());
+  }
+
+  template <typename T>
+  Status Visit(const T&,
+               typename internal::DictionaryTraits<T>::MemoTableType* = nullptr) {
+    using ArrayType = typename TypeTraits<T>::ArrayType;
+    using DictTraits = typename internal::DictionaryTraits<T>;
+    using MemoTableType = typename DictTraits::MemoTableType;
+
+    MemoTableType memo_table;
+    if (out_transpose_maps_ != nullptr) {
+      out_transpose_maps_->clear();
+      out_transpose_maps_->reserve(types_.size());
+    }
+    // Build up the unified dictionary values and the transpose maps
+    for (const auto& type : types_) {
+      const ArrayType& values = checked_cast<const ArrayType&>(*type->dictionary());
+      if (out_transpose_maps_ != nullptr) {
+        std::vector<int32_t> transpose_map;
+        transpose_map.reserve(values.length());
+        for (int64_t i = 0; i < values.length(); ++i) {
+          int32_t dict_index = memo_table.GetOrInsert(values.GetView(i));
+          transpose_map.push_back(dict_index);
+        }
+        out_transpose_maps_->push_back(std::move(transpose_map));
+      } else {
+        for (int64_t i = 0; i < values.length(); ++i) {
+          memo_table.GetOrInsert(values.GetView(i));
+        }
+      }
+    }
+    // Build unified dictionary array
+    std::shared_ptr<ArrayData> data;
+    RETURN_NOT_OK(DictTraits::GetDictionaryArrayData(pool_, value_type_, memo_table,
+                                                     0 /* start_offset */, &data));
+    *out_values_ = MakeArray(data);
+    return Status::OK();
+  }
+};
+
+Status DictionaryType::Unify(MemoryPool* pool, const std::vector<const DataType*>& types,
+                             std::shared_ptr<DataType>* out_type,
+                             std::vector<std::vector<int32_t>>* out_transpose_maps) {
```

Comment:
We should also consider the case where we want to conform a dictionary to an other, without unification. For example:

```
[a, b, c, a, b, c], dictionary [a, b, c]

conform to dictionary [a, c, d]

yields

[a, null, c, a, null, c]
```

This isn't unification so much as conformance. This was what was meant by my comment "Conform the dictionary indices to reference a common dictionary" in ARROW-554. The use case would be implementing casts between dictionary types -- this would include a safe and unsafe mode (safe would error if a value is made null)

Since this use case is not addressed in this patch (unless I missed it), let us create a follow up JIRA

Reply:
Hmm, I see. I have misinterpreted the original JIRA then. In any case, I was thinking of the CSV use case, where we'll be unifying dictionaries at the end.

Reply:
I opened https://issues.apache.org/jira/browse/ARROW-4097